File: plugin/download/ios/Classes/DownloadPlugin.swift
import Flutter
import UIKit

public class DownloadPlugin: NSObject, FlutterPlugin {
    
    private let downloadService: DownloadService = DownloadServiceImpl(downloadRepository: DownloadRepositoryImpl())
    
    private static let METHOD_CHANNEL = "http_downloader/download"
    private static let PROGRESS_EVENT_CHANNNEL = "http_downloader/download/progress"
    private static let FINISH_EVENT_CHANNEL = "http_downloader/download/finished"
    
    static var progressEventSink: FlutterEventSink? = nil
    static var finishEventSink: FlutterEventSink? = nil
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(name: METHOD_CHANNEL, binaryMessenger: registrar.messenger())
        let progressEventChannel = FlutterEventChannel(name: PROGRESS_EVENT_CHANNNEL, binaryMessenger:  registrar.messenger())
        let finishEventChannel = FlutterEventChannel(name: FINISH_EVENT_CHANNEL, binaryMessenger:  registrar.messenger())
        
        progressEventChannel.setStreamHandler(ProgressStreamHandler())
        finishEventChannel.setStreamHandler(FinishStreamHandler())
      
      
        let instance = DownloadPlugin()
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        NSLog(call.method)
        if call.method == "getPlatformVersion" {
            result("iOS: \(UIDevice.current.systemVersion)")
            return
        }
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError.init(code: "errorSetDebug", message: "data or format error", details: nil))
            return
        }
        
        
      switch(call.method) {
      case "createDownload":
          
          guard let urlString = args["urlString"] as? String else {
              result(FlutterError.init(code: "errorSetDebug", message: "data or format error", details: nil))
              return
          }
          guard let isConcurrent = args["isConcurrent"] as? Bool else {
              result(FlutterError.init(code: "errorSetDebug", message: "data or format error", details: nil))
              return
          }
          
          Task {
              let serviceResult: ServiceResult<Int> = await self.downloadService.createDownload(urlString: urlString, isConcurrent:  isConcurrent)
              let response = MethodChannelResponse(serviceResult: serviceResult)
              result(String(data: try JSONEncoder().encode(response), encoding: .utf8) )
          }
              
              
          
                  
      case "resumeDownload":
          guard let downloadID = args["downloadID"] as? String else {
              result(FlutterError.init(code: "errorSetDebug", message: "data or format error", details: nil))
              return
          }
          
          Task {
              do {
                  let serviceResult: ServiceResult<Int> = await self.downloadService.resumeDownload(downloadID: downloadID)
                  let response = MethodChannelResponse(serviceResult: serviceResult)
                  result(String(data: try JSONEncoder().encode(response), encoding: .utf8))
              } catch {
                  result(FlutterError.init(code: "jsonSerailizationError", message: "resumeDownload response serialization failed", details: nil))
              }
          }
          
          
          
      case "cancelDownload":
          guard let downloadID = args["downloadID"] as? String else {
              result(FlutterError.init(code: "errorSetDebug", message: "data or format error", details: nil))
              return
          }
          
          Task {
              do {
                  let serviceResult: ServiceResult<Int> = await self.downloadService.cancelDownload(downloadID: downloadID)
                  let response = MethodChannelResponse(serviceResult: serviceResult)
                  result(String(data: try JSONEncoder().encode(response), encoding: .utf8))
              } catch {
                  result(FlutterError.init(code: "jsonSerailizationError", message: "cancelDownload response serialization failed", details: nil))
              }
          }
      case "pauseDownload":
          guard let downloadID = args["downloadID"] as? String else {
              result(FlutterError.init(code: "errorSetDebug", message: "data or format error", details: nil))
              return
          }
          
          Task {
              do {
                  let serviceResult: ServiceResult<Int> = await self.downloadService.pauseDownload(downloadID: downloadID)
                  let response = MethodChannelResponse(serviceResult: serviceResult)
                  result(String(data: try JSONEncoder().encode(response), encoding: .utf8))
              } catch {
                  result(FlutterError.init(code: "jsonSerailizationError", message: "pauseDownload response serialization failed", details: nil))
              }
          }
      case "manualPauseDownload":
          guard let downloadID = args["downloadID"] as? String else {
              result(FlutterError.init(code: "errorSetDebug", message: "data or format error", details: nil))
              return
          }
          
          Task {
              do {
                  let serviceResult: ServiceResult<Int> = await self.downloadService.manualPauseDownload(downloadID: downloadID)
                  let response = MethodChannelResponse(serviceResult: serviceResult)
                  result(String(data: try JSONEncoder().encode(response), encoding: .utf8))
              } catch {
                  result(FlutterError.init(code: "jsonSerailizationError", message: "manualPauseDownload response serialization failed", details: nil))
              }
          }
      default:
          NSLog(call.method)
          result(FlutterMethodNotImplemented)
      }
    }
}

class ProgressStreamHandler: NSObject, FlutterStreamHandler {
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        NSLog("ProgressStreamHandler Configured")
        DownloadPlugin.progressEventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
}

class FinishStreamHandler: NSObject, FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        NSLog("FinishStreamHandler Configured")
        DownloadPlugin.finishEventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
}

File: plugin/download/android/src/main/kotlin/com/example/perfect_corp_homework/download/DownloadPlugin.kt
package com.example.perfect_corp_homework.download

import android.util.Log
import com.example.perfect_corp_homework.download.api.MethodChannelResponse
import com.example.perfect_corp_homework.download.model.DownloadEntity
import com.example.perfect_corp_homework.download.repository.DownloadRepositoryImpl
import com.example.perfect_corp_homework.download.service.DownloadService
import com.example.perfect_corp_homework.download.service.DownloadServiceImpl
import com.google.gson.Gson

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import org.slf4j.LoggerFactory

/** DownloadPlugin */
class DownloadPlugin: FlutterPlugin, MethodCallHandler {

  companion object {
    private val METHOD_CHANNEL = "http_downloader/download"
    private val PROGRESS_EVENT_CHANNNEL = "http_downloader/download/progress"
    private val FINISH_EVENT_CHANNEL = "http_downloader/download/finished"

    private val downloadService: DownloadService = DownloadServiceImpl(DownloadRepositoryImpl())
    private val log = LoggerFactory.getLogger("MainActivity")

    private var progressEventSink: EventChannel.EventSink? = null
    var finishedEventSink: EventChannel.EventSink? = null


    val gson = Gson()

    val sendProgressEvent = { downloads: List<DownloadEntity> ->
      progressEventSink!!.success(gson.toJson(downloads))
    }

    val sendFinishedEvent = { downloadEntity: DownloadEntity ->
      finishedEventSink!!.success(gson.toJson(downloadEntity))
    }
  }

  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, METHOD_CHANNEL)
    channel.setMethodCallHandler(this)


    var progressEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, PROGRESS_EVENT_CHANNNEL).setStreamHandler(
      object: EventChannel.StreamHandler{
        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
          log.info("progressEventChannel Constructed")
          progressEventSink = events
        }

        override fun onCancel(arguments: Any?) {
          log.info("progressEvent Canceled")
          progressEventSink = null
        }

      }
    )

    var finishedEventChannel = EventChannel(flutterPluginBinding.binaryMessenger, FINISH_EVENT_CHANNEL).setStreamHandler(
      object : EventChannel.StreamHandler {
        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
          log.info("finishedEventChannel Constructed")
          finishedEventSink = events
        }

        override fun onCancel(arguments: Any?) {
          log.info("finishedEventChannel Canceled")
          finishedEventSink = null
        }
      }
    )
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when(call.method) {
      "createDownload" -> runBlocking {
        val urlString: String = call.argument<String>("urlString") as String
        val isConcurrent: Boolean = call.argument<Boolean>("isConcurrent") as Boolean

        launch(Dispatchers.Default) {

          val serviceResult = downloadService.createDownload(urlString = urlString, isConcurrent = isConcurrent)

          val methodChannelResponse = MethodChannelResponse<Int>(serviceResult)

          val jsonString = gson.toJson(methodChannelResponse)

          result.success(jsonString)
        }
      }
      "pauseDownload" -> runBlocking {
        launch(Dispatchers.Main) {
          val downloadID: String = call.argument<String>("downloadID") as String
          val serviceResult = downloadService.pauseDownload(downloadID = downloadID)
          val methodChannelResponse = MethodChannelResponse<Int>(serviceResult)
          val jsonString = gson.toJson(methodChannelResponse)
          result.success(jsonString)
        }
      }
      "resumeDownload" -> runBlocking {
        launch(Dispatchers.Main) {
          val downloadID: String = call.argument<String>("downloadID") as String
          val serviceResult = downloadService.resumeDownload(downloadID = downloadID)
          val methodChannelResponse = MethodChannelResponse<Int>(serviceResult)
          val jsonString = gson.toJson(methodChannelResponse)
          result.success(jsonString)
        }
      }
      "cancelDownload" -> runBlocking {
        launch(Dispatchers.Main) {
        val downloadID: String = call.argument<String>("downloadID") as String
        val serviceResult = downloadService.cancelDownload(downloadID = downloadID)
        val methodChannelResponse = MethodChannelResponse<Int>(serviceResult)
        val jsonString = gson.toJson(methodChannelResponse)
        result.success(jsonString)
        }
      }
      "manualPauseDownload" -> runBlocking {
        launch(Dispatchers.IO) {
          val downloadID: String = call.argument<String>("downloadID") as String
          val serviceResult = downloadService.pauseDownload(downloadID = downloadID)
          val methodChannelResponse = MethodChannelResponse<Int>(serviceResult)
          val jsonString = gson.toJson(methodChannelResponse)
          result.success(jsonString)
        }
      }
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}
